File: Cpp/LeetCode/103.cpp
//
// Created by caizixiang on 2019-07-29.
//

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

 //Definition for a binary tree node.
 struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> order;
        if (!root)
            return order;

        int flag = 1;
        stack<TreeNode*> current_layer;
        current_layer.push(root);

        while (!current_layer.empty())
        {
            vector<int> this_layer;
            stack<TreeNode*> next_layer;
            while (!current_layer.empty())
            {
                TreeNode* node = current_layer.top();
                current_layer.pop();
                this_layer.push_back(node->val);
                if (flag)
                {
                    if (node->left)
                        next_layer.push(node->left);
                    if (node->right)
                        next_layer.push(node->right);
                } else
                {
                    if (node->right)
                        next_layer.push(node->right);
                    if (node->left)
                        next_layer.push(node->left);
                }

            }

            order.push_back(this_layer);
            current_layer = next_layer;
            flag = 1 - flag;
        }

        return order;
    }
};

File: Python/LeetCode/103.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        stack = [[root], []]
        flag = 0
        order = []
        while stack[flag] != [None]*len(stack[flag]):
            this_layer = []
            while stack[flag]:
                node = stack[flag].pop()
                if node:
                    this_layer.append(node.val)
                    if flag:
                        stack[1-flag] += [node.right, node.left]
                    else:
                        stack[1-flag] += [node.left, node.right]

            stack[flag] = []
            flag = 1 - flag
            order.append(this_layer)

        return order



